On-demand policy wildcarding

jrajahalme opened the pull request:
Replace the current model of mass L7 wildcarding with on-demand wildcarding to allow for redirecting some but not all selectors for a given L4 filter.

L4 filters without L7 rules are now wildcarded either when merged with L4/L7 rules, or when used by the L7 proxies.

Bpf policy lookup now always looks up both L3/L4 and L4-only rules to find out if either of them redirects to proxy. This allows elimination of additionally generated policy map entries that would otherwise needed to ensure L4-only policy redirects to the proxy in the presence of L3/L4 policies that do not redirect.

Some context for reviewing:
- The functionality of `L4Filter.CachedSelectors` has been merged into the `L4Filter.L7RulesPerEp` (which has been renamed as `L7RulesPerSelector`). This is needed to efficiently keep track of which selectors have L7 rules and which not. The latter case is represented by a nil value in the map.
- L7 proxy code dealing with `L7RulesPerEp` has been modified to threat `nil` values as L7 wildcards (DNS, Kafka, Envoy)
- L7 wildcarding rules are generated on-demand when merging a non-L7 L4 rule with a L4/L7 rule.
- `L4Filter.ToKeys()` has been renamed as `L4Filter.ToMapState` as it now returns a map (MapState) instead of just the keys. Map entries tell if the entry needs L7 redirection or not. This makes it possible for a single L4 filter to create redirects for some of the allowed security IDs, but not necessarily all of them.
- Tracking of MapState entries now relies on each CachedSelector tracked by an L4Filter to be in the L7RulesPerSelector map. nil vs. non-nil value in the map is used to determine if the new map entry needs redirection or not. This is then used to test if the proxy port value needs to be set, so that even if the L7Proxy type of the L4 filter is set, selectors without L7 rules do not generate proxy redirection in the bpf policy map.
- These changes mean that L3-dependent L7 processing in the proxy now relies on L3/L4 entries in addition to L3/L4/L7 entries in the NetworkPolicy. This required removal of some sidecar-specific code from `cilium-envoy` that was never exercised earlier due to the "mass wildcarding" model.
- `policy/selectorcache.go` required an interesting fix relating to use of interfaces as map keys (like in `L4Filter.L7RulesPerSelector` :-) 
- Policy map lookup order in the bpf datapath is changed from `L3/L4 -> L3/* -> */L4 -> */*` to `L3/L4 -> */L4 -> L3/* -> */*` as this removes the need to generate L7 wildcards for L3-only entries.
- L3/L4 keys are now supressed from the bpf policy maps if a L4-only key exists and does what L3/L4 entry would have needed to do. This removes the need to generate wildcards for L3/L4 keys when L4-only key redirects.

Fixes: #9996

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10054)
<!-- Reviewable:end -->

joestringer commented on daemon/policy_test.go:
```diff
@@ -364,6 +422,90 @@ func (ds *DaemonSuite) TestUpdateConsumerMap(c *C) {
+				Port:     80,
+				Protocol: envoy_api_v2_core.SocketAddress_TCP,
+				Rules: []*cilium.PortNetworkPolicyRule{
+					{},
```
Does this empty brackets denote L4 allow-all without rules?

jrajahalme replied:
It denotes L4 allow-all without any L3 restrictions (nil Remotepolicies as well as nil L7)

joestringer commented on pkg/envoy/server.go:
```diff
@@ -831,6 +836,48 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, l7Parser policy.L7Parse
+		}
+
+		if l7 != nil {
+			log.Warning("Wildcard L4 L7DataMap has L7 rules!")
```
Can we get more info here? From my reading this is basically not possible to hit right now but if someone ever did hit it, it might be useful to have some kind of selector info to help debug?

jrajahalme replied:
Will add the selector and the l7 rules to the message.

joestringer commented on pkg/envoy/server.go:
```diff
@@ -855,32 +902,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
-			}
+
+		if l4.Port == 0 {
+			// Wildcard L4 rule, must generate L7 allow-all in case there are other
```
nit: 
```suggestion
			// L3-only rule, must generate L7 allow-all in case there are other
```

I feel like we've always talked about l3-only, l4-only, l3-dependent-l7 (L3+L4+L7), l4-dependent-l7 (L4+L7) as the standard way to think about these rules.

jrajahalme replied:
Will change.

raybejjani reviewed (approved):
One straightforward change to the dns tests (sent offline) but LGTM otherwise.

One thing I noticed was that this policy resolved to a L7 wildcard in the proxy. It's "correct" but master handled it as if the L3/L4 rule didn't exist. This is exactly the point of this PR, right?
```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: "tofqdn-dns-visibility"
spec:
  endpointSelector:
    matchLabels:
      any:org: alliance
  egress:
  - toEndpoints:
    - matchLabels:
       "k8s:io.kubernetes.pod.namespace": kube-system
       "k8s:k8s-app": kube-dns
    toPorts:
      - ports:
         - port: "53"
           protocol: ANY

  - toEndpoints:
    - matchLabels:
       "k8s:io.kubernetes.pod.namespace": kube-system
       "k8s:k8s-app": kube-dns
    toPorts:
      - ports:
         - port: "53"
           protocol: ANY
        rules:
          dns:
            - matchPattern: "google.com"
```

joestringer commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -550,4 +550,32 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
 	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
 	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
+
+	// | EP3  | DstID1 |      53 | *    |
```
Please update the table at the start of this function to describe this case in more detail, so it's easier to get a summary of how this compares to the other test cases here.

joestringer replied:
Actually, this line doesn't make sense in context of the current batch of tests at all. Similarly on line 571.

joestringer replied:
Can we just expand the set of tests to include all the cases at the start to avoid this confusion? Or are you specifically trying to test incremental policy, in which case would it help to have a separate test for incremental testing?

joestringer replied:
I guess this may have been addressed already by Ray's patch.

raybejjani replied:
yup, that was exactly my request. It should declare everything up top now and only try the new cases at the end.

joestringer commented on pkg/policy/distillery_test.go:
```diff
@@ -424,26 +420,24 @@ func Test_MergeRules(t *testing.T) {
 		{4, api.Rules{rule____NoAllow, rule____NoAllow, ruleL3L4__Allow, rule____NoAllow, rule____NoAllow}, MapState{mapKeyAllowFooL4: mapEntryL7None_}},
 		{5, api.Rules{rule____NoAllow, rule____NoAllow, ruleL3L4__Allow, rule____NoAllow, ruleL3____Allow}, MapState{mapKeyAllowFooL4: mapEntryL7None_, mapKeyAllowFoo__: mapEntryL7None_}},
-		{6, api.Rules{rule____NoAllow, rule____NoAllow, ruleL3L4__Allow, rule__L4__Allow, rule____NoAllow}, MapState{mapKeyAllow___L4: mapEntryL7None_}},                                    // Differs from spreadsheet(!)
+		{6, api.Rules{rule____NoAllow, rule____NoAllow, ruleL3L4__Allow, rule__L4__Allow, rule____NoAllow}, MapState{mapKeyAllowFooL4: mapEntryL7None_, mapKeyAllow___L4: mapEntryL7None_}}, // Differs from spreadsheet(!)??
```
In this particular case it doesn't differ from the spreadsheet any more with this change. Would you mind double-checking the rest of these distillery test changes against the spreadsheet?

https://docs.google.com/spreadsheets/d/1WANIoZGB48nryylQjjOw6lKjI80eVgPShrdMTMalLEw/edit#gid=0

jrajahalme replied:
Checked the changed lines against the spreadsheet. The cases 12-15 all have an L4-only rule redirecting to proxy, and an L3/L4 rule that is allowed without redirecting. Now the latter case will not be sent to the proxy as it is not required for policy enforcement, but this may be problematic for a visibility policy?

jrajahalme replied:
Updated the comments to reflect the above!

joestringer commented on pkg/policy/l4_filter_test.go:
```diff
@@ -243,22 +249,15 @@ func (ds *PolicyTestSuite) TestMergeAllowAllL3AndShadowedL7(c *C) {
-		L7Parser:        "http",
-		L7RulesPerEp: L7DataMap{
-			wildcardCachedSelector: &PerEpData{
+		Port:     80,
```
Was the previous comment above this misleading or wrong? It seems like it's trying to provide additional guidance on how to interpret the testing operations below.

jrajahalme replied:
Now that the wildcard rule is explicitly generated, the explanation was not entirely correct anymore, and IMO not needed.

jrajahalme replied:
Note: I don't know where the L7 wildcard would have been created for the proxy before this change. Maybe it worked due to L4-only allow possibly NOT redirecting to the proxy at all?

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -125,10 +125,10 @@ func (p *selectorPolicy) DistillPolicy(policyOwner PolicyOwner) *EndpointPolicy
 	//   is already added to the PolicyMapState
 	//
-	// - PolicyMapChanges may congtain a deletion of an entry that
+	// - policyMapChanges may congtain a deletion of an entry that
```
```suggestion
	// - policyMapChanges may contain a deletion of an entry that
```

jrajahalme replied:
Fixed, thanks!

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -154,26 +154,33 @@ func (p *EndpointPolicy) computeDesiredL4PolicyMapEntries() {
+func (p *EndpointPolicy) ConsumeMapChanges() (adds, deletes MapState) {
+	p.selectorPolicy.SelectorCache.mutex.Lock()
+	defer p.selectorPolicy.SelectorCache.mutex.Unlock()
+	return p.policyMapChanges.consumeMapChanges()
```
Is this a potential bugfix?

jrajahalme replied:
Not sure. It is needed now, but likely not if all L4 entries redirected before if there was an L7 on the same port.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -319,9 +309,10 @@ func (l4 *L4Filter) IdentitySelectionUpdated(selector CachedSelector, selections
-	// This logic mirrors the one in ToKeys().
-	if l4.AllowsAllAtL3() && !l4.HasL3DependentL7Rules() {
+	// Skip updates on wildcard selectors, as datapath and L7
+	// proxies do not need enumeration of all ids for L3 wildcard.
```
I guess the implication here is that if L3 is wildcarded then we already calculated the L4-only / L4-dependent L7 changes necessary for the datapath, so the fact that some new identity showed up for an L3 doesn't make any difference to the eventual policy.

jrajahalme replied:
Right. We generate the L4-only key (L3=0) regardless the actual IDs the wildcard selector maps to.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -373,20 +365,24 @@ func (l7 L7DataMap) GetRelevantRulesForKafka(nid identity.NumericIdentity) []api
 	for cs, r := range l7 {
 		if cs.IsWildcard() || cs.Selects(nid) {
+			if r == nil {
+				r = &PerSelectorPolicy{L7Rules: api.L7Rules{Kafka: []api.PortRuleKafka{}}}
```
nit: We should probably standardize these kinds of `AllowAllKafka` representations into public variables to make it more clear and ensure consistency across different parts of the code.

joestringer replied:
nit2: Seems like we sanitized the rule used here before, not sure if there was a specific reason that was necessary..?

jrajahalme replied:
Could also move to make Kafka proxy accept `nil` instead?

IMO Sanitizing would not change this rule. egress Kafka is not supported, but here the call comes from the already running proxy.

joestringer replied:
Well, it's a debate with Kafka whether to make any further changes given the Envoy support for that protocol now..

joestringer commented on pkg/policy/l4.go:
```diff
@@ -485,32 +479,44 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
+func (l4 *L4Filter) removeSelectors(selectorCache *SelectorCache, all bool) {
+	selectors := make(CachedSelectorSlice, 0, len(l4.L7RulesPerSelector))
+	for cs, l7 := range l4.L7RulesPerSelector {
+		if all || l7 == nil {
```
This function adds a bool for `all` but AFAICT the `false` version is never used. I also can't quite wrap my head around why we don't release references in the selectorCache for selectors used with L7 rules assuming that we ever passed false. Do we need this extra complexity?

jrajahalme replied:
This is a remnant from an intermediate state of the code, will clean it up!

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -67,15 +67,40 @@ type MapStateEntry struct {
 	// If 0 (default), there is no proxy redirection for the corresponding
 	// Key.
+	// Values less than 1024 may be used for internal purposes, but are never used
+	// as actual proxy listening port numbers.
```
Is there any code that ensures this constraint?

I guess if we're only using 1, then it's unlikely we're going to cause problems with... *\*checks notes\**... the deprecated TCPMUX protocol? ;-)

jrajahalme replied:
All callers compare the entry against 0 value, so the 1 value is actually not used for comparisons. Will enforce this by un-exporting the `RedirectEntry`.

jrajahalme replied:
Also changed the comment to indicate that any non-zero value signifies a redirect.

joestringer commented on pkg/policy/repository_test.go:
```diff
@@ -1280,54 +1245,50 @@ func (ds *PolicyTestSuite) TestWildcardL4RulesEgress(c *C) {
 	c.Assert(err, IsNil)
 
+	// Bar1 should not be forwarded to the proxy, but if it is (e.g., for visibility),
+	// the L3/L4 allow should pass it without a L7 wildcard.
```
nit:
```suggestion
	// the L3/L4 allow should pass it without an explicit L7 wildcard.
```

jrajahalme replied:
Done.

joestringer commented on pkg/policy/rule_test.go:
```diff
@@ -1962,12 +1973,27 @@ func (ds *PolicyTestSuite) TestL3L4L7Merge(c *C) {
+					HTTP: []api.PortRuleHTTP{{Path: "/", Method: "GET"}},
+				},
+			},
+			cachedSelectorC: nil,
```
nit: Looks like you're specifying `cachedSelectorC` here twice?

jrajahalme replied:
Not needed, will remove the other!

joestringer commented on test/runtime/Policies.go:
```diff
@@ -528,31 +532,17 @@ var _ = Describe("RuntimePolicies", func() {
+		// Since policy allows connectivity on L3 from app3 to httpd2, and such
+		// packets are not forwarded to the proxy, we expect no changes in proxy
+		// stats.
+		checkProxyStatistics(app3EndpointID, 2, 4, 2, 2, 2)
```
:tada:

joestringer commented on pkg/policy/rule.go:
```diff
@@ -58,46 +58,44 @@ func (r *rule) String() string {
-			// Update selector owner to the existingFilter
-			selectorCache.ChangeUser(cs, from, l4)
+func (epd *PerSelectorPolicy) appendL7WildcardRules(ctx *SearchContext) *PerSelectorPolicy {
+	switch {
```
nits: what about the default case, including go extensions?

joestringer replied:
Looking at the other L7 handling piece in this file, I guess we don't support merging for such L7 rules. That said I think it's probably worth raising at least a policy trace line, but probably also a warning of some kind that you're doing something not-supported in that case?

jrajahalme replied:
go extensions do not need a L7 rule to be triggered as the parser type/id is explicitly invoked in the policy. Even if no rules acts as a wildcard, an empty set would need to be appended here. Will fix!

jrajahalme replied:
Added a commit to address this :-)

joestringer reviewed (approved):
Mostly nits and questions to make sure I'm following correctly below, and a couple of notes where we might be able to help subsequent readers of the code to understand.

We should double-check the google doc corresponding to distillery_test, at least the first case now conforms with the doc.

Also some of the new L7 rules handling statements don't handle default case in any way, which I guess results in silent ignorance of things that the current code wasn't thought through for. Might be good to add some kind of message to those cases?

None of the above are major hold backs. Only other remaining question I have is whether we've been able to run this in non-trivial situations and observe things like the selectorCache to ensure that we got the reference counting correct.

jrajahalme committed:
policy: Add on-demand wildcarding for generic L7

Generic L7 also needs to be wildcarded with an empty key-value map if
an L3/L4 rule is merged with a generic L7 rule.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add wildcard member to L4Filter

Keep track if the L4Filter has a wildcard (cached) selector.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Skip generating L3/L4 keys in some cases

Skip generating L3/L4 keys if L4-only key (for the same L4 port and
protocol) has the same effect w.r.t. redirecting to the proxy or not,
considering that L3/L4 key should redirect if L4-only key does. In
summary, if have both L3/L4 and L4-only keys:

L3/L4        L4-only      Skip generating L3/L4 key
redirect     no redirect  no (redirection is not needed for the L4 port in general)
no redirect  no redirect  yes (same entry in both keys)
redirect     redirect     yes (same entry in both keys)
no redirect  redirect     yes (redirection is needed for the L4 port)

Signed-off-by: Jarno Rajahlame <jarno@covalent.io>

joestringer commented on pkg/policy/l4.go:
```diff
@@ -185,6 +185,8 @@ type L4Filter struct {
 	Protocol api.L4Proto `json:"protocol"`
 	// U8Proto is the Protocol in numeric format, or 0 for NONE
 	U8Proto u8proto.U8proto `json:"-"`
+	// wildcard is the cached selector representing a wildcard in this filter, if any.
```
Might be helpful for future developers to note what this means in terms of higher-layer policies, which cases this wildcard field would be set, etc.

jrajahalme replied:
```
	// wildcard is the cached selector representing a wildcard in this filter, if any,
	// This is nil when none of the selectors in 'L7RulesPerSelector' is a wildcard.
	// When there is a wildcard selector in 'L7RulesPerSelector' this is set to that
	// same selector, which can then be used as a map key to find the correspondng L7
	// policy.
```

joestringer commented on pkg/policy/l4.go:
```diff
@@ -256,7 +256,28 @@ func (l4 *L4Filter) ToMapState(direction trafficdirection.TrafficDirection) MapS
+		// no redirect  redirect     yes  (must redirect if L4-only redirects)
+		//
+		// have wildcard?        this is a L3L4 key?  not the "no" case?
+		if l4.wildcard != nil && cs != l4.wildcard && !(l7 != nil && wildcardL7Policy == nil) {
```
Is it possible to demonstrate this specific line of functionality concisely with a unit test as well?

There's a couple of other cases not mentioned in the table immediately above, which is the presence (or lack thereof) of an l4-only rule at all, which is what I think the first term of the conditional statement here is checking.

Also want to make sure that this is handling correctly while iterating the `l4.L7RulesPerSelector` entry for the L4-only filter case. (EDIT: Ah, that's the `cs != l4.wildcard` case, OK).

jrajahalme replied:
The table driven cases of `distillery_test.go` cover this line. I started with the 'no' lines for non-existing L4-only cases, but took it away. I'll put them back :-)

joestringer reviewed (changes requested):
Couple of other minor comments on the new pieces.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -256,7 +256,28 @@ func (l4 *L4Filter) ToMapState(direction trafficdirection.TrafficDirection) MapS
+		//
+		// have wildcard?        this is a L3L4 key?  not the "no" case?
+		if l4.wildcard != nil && cs != l4.wildcard && !(l7 != nil && wildcardL7Policy == nil) {
+			continue
```
Is there an appropriate way to log the skipping here? I'm concerned that if this silently discards some map entry generation then in future we'll be wondering where it went.

jrajahalme replied:
Good point, I'll add a debug line. Will change "ToKeys" to "ToMapState" also in the debug messages.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -256,7 +256,28 @@ func (l4 *L4Filter) ToMapState(direction trafficdirection.TrafficDirection) MapS
+		//
+		// L3/L4        L4-only      Skip generating L3/L4 key
+		// redirect     no redirect  no   (this case tested below)
+		// no redirect  no redirect  yes  (same effect)
```
Does `l4.L7RulesPerSelector` have an entry for these cases where there is no redirect? ie are we actually handling that case in this loop?

joestringer replied:
Technically I think we could still install the l3/l4 key in this case and it would save one map lookup for packets that match the l3/l4 policy on the dataplane, but we can probably optimize this later.

jrajahalme replied:
As one of the changes in this PR overall, `l4.L7RulesPerSelector` now has an entry also for non-redirects, with `nil` value.

On the second point, yes we can do that, but the effect is now smaller due to the change in the lookup order in the datapath.

joestringer commented on pkg/policy/distillery_test.go:
```diff
@@ -420,16 +385,16 @@ func Test_MergeRules(t *testing.T) {
-		{13, api.Rules{rule____NoAllow, rule__L4L7Allow, ruleL3L4__Allow, rule____NoAllow, ruleL3____Allow}, MapState{mapKeyAllow___L4: mapEntryL7Proxy, mapKeyAllowFooL4: mapEntryL7None_, mapKeyAllowFoo__: mapEntryL7None_}}, // Differs from spreadsheet(!) L3L4 does not redirect
-		{14, api.Rules{rule____NoAllow, rule__L4L7Allow, ruleL3L4__Allow, rule__L4__Allow, rule____NoAllow}, MapState{mapKeyAllowFooL4: mapEntryL7None_, mapKeyAllow___L4: mapEntryL7Proxy}},                                    // Differs from spreadsheet(!), L3L4 not going to proxy (not required, but maybe problematic for a visibility policy?)
-		{15, api.Rules{rule____NoAllow, rule__L4L7Allow, ruleL3L4__Allow, rule__L4__Allow, ruleL3____Allow}, MapState{mapKeyAllow___L4: mapEntryL7Proxy, mapKeyAllowFoo__: mapEntryL7None_, mapKeyAllowFooL4: mapEntryL7None_}}, // Differs from spreadsheet(!) L3L4 does not redirect
+		{12, api.Rules{rule____NoAllow, rule__L4L7Allow, ruleL3L4__Allow, rule____NoAllow, rule____NoAllow}, MapState{mapKeyAllow___L4: mapEntryL7Proxy}},                                    // identical L3L4 entry suppressed
```
The comment for these ones says `identical`, but this case is the L4-L7 rule overriding the L3+L4(no-L7).

jrajahalme replied:
Historical reasons :-) I used to 1st convert a non-redirect case to a redirect case when needed, and then remove redundant L3L4 cases. These are now done in one step. I'll update the comment(s).

joestringer reviewed (approved):
Few more minor nits, nothing consequential. Consider following up separately afterwards on these points.

jrajahalme committed:
Address review comments (round 2)

Addressing Joe's review comments.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
ginkgo: Bump all timeouts

Have been hit by back to back provisioning failures due to timeouts
being hit on processes that seem to have been progressing, albeit
maybe slower than normal.

After timeout VM provisioning is tried again, but overall we seem to
be using more time due to cutting progressing processes off too
eagerly.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Fix test combining of L4 and L7.

Specify slice size as 0 when making it so that it is not full of zero
entries to begin with.

This has no effect on test results, but fixing this reduces confusion
in reading the test output.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Unify CachedSelectors and L7RulesPerEp

Currently all selectors used by a policy are stored in
'CachedSelectors', while a subset of them having L7 rules are present
in the L7RulesPerEp. Moving to an on-demand L7 wildcarding sceme
requires keeping track of selectors without L7 rules (as those need
wildcarding if merged with a filter with L7 rules). Do this by placing
all selectors into L7RulesPerEp, while renaming it to
'L7RulesPerSelector'. Selectors without L7 rules have a value 'nil' in
this map, which now represents "L7 wildcard". Remove 'CachedSelectors'
as it is now fully redundant.

Proxies (DNS, Kafka, Envoy) are updated to accept 'nil' as a wildcard
rule.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Always pass CachedSelectors as the implmenation type.

The CachedSelector interface values passed from the selector cache are
used as map keys in the policy implementation. To properly function as
map keys, they must always be passed as the pointer to the
implementation of the same type.

'notifyUsers' was implemented by the "base class" (selectorManager),
so when it passed pointer to itself as a CachedSelector, the type
inside the interface pointed to the selectorManager type, not the
original fqdnSelector or labelIdentitySelector, causing map lookup
failures.

The CachedSelector passed by 'notifyUsers' was not previously used as
a map key, so this error did not surface earlier.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Add L3-dependent L7 policy test

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Change ToKeys() to ToMapState()

Rename ToKeys() to ToMapState() and return a map instead just the
keys. A map value with 0 port number means no redirection, while a map
value with any other port number signifies proxy redirection. The port
value returned (for entires that need proxy redirection) will be
replaced by the actual proxy listening port numbers by the caller of
ToMapState().

This allows different bpf policy map entiries derived from the same L4
filter redirect to the proxy or not. To help separate these two cases
we have to remove the mass wildcarding of policy rules by removing the
calls to wildcardL3L4Rule() and wildcardL3L4Rules().

These changes revealed a policy evaluation bug in cilium-envoy. Use a new
Envoy image to address this. Note that this bug does not trigger with the
provious "mass wildcarding" behavior of cilium-agent.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Rename PerEpData to PerSelectorPolicy

This data has never been per Endpoint, renaming should clarify this.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Remove AllowsAllAtL3() and HasL3DependentL7Rules()

These are no longer needed so remove them.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Make FQDN runtime tests more robust.

Allow docker network "world" to exist in advance. This is helpful when
focus testing different test on the same VM.

Do not expect FQDN selectors to have selected any IDs right after the
policy is first imported. It is common to have no IDs when the policy
is first imported, but no IPs have yet been resoved via the DNS proxy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Add test with multiple ingress rules on the same port

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

raybejjani committed:
Apply all full-path-dependence rules at the same time.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

jrajahalme committed:
Address review comments (round 1)

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Give priotity for L7 redirect entries

Two selectors can select the same security identity, and the policy
can specify L7-rules for one of them, but not the other. While L7
enforcement would make this policy inconsistent, such a policy is
valid for L7 visibility. To ensure consistent behavior give priority
to redirect in the case the same map key has both a redirect and
non-redirect.

The implementation requires updates on a given security ID to happen
atomically. To achive this ConsumeMapChanges() is moved to the
endpoint policy, which can lock the selectorcache to guarantee that
concurrent updates on any new security ID have completed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Look up L4 before L3 and redirect to proxy if either L4 result redirects

Always look up both L3/L4 and L4-only to find out if either of them
redirects to proxy. This allows elimination of additionally generated
policy map entries that would otherwise needed to ensure L4-only
policy redirects to the proxy in the presence of L3/L4 policies that
do not redirect.

This implementation is somewhat constrained by the bpf verifier not
being able to handle two policy lookup results at the same time. This
is why the packet is accounted towards both rules if it matches both
an L3/L4 rule that does not redirect and an L4-only rule. This
double-accounting could be eliminated by repeating the first policy
map lookup, but that would add otherwise unnecessary overhead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tgraf reviewed (approved):
I've reviewed this PR in terms of risk merging this late and after longer evaluation I'm OK to merge it. I think it's borderline and we should avoid any such invasive PR this late in the cycle but the alternative is not much better given we expect this to become a major use case for Cilium.
